link/perf_event.go

```diff
@@ -67,10 +67,86 @@ const (
 	uretprobeEvent
 )
 
+// perfEventLink wraps perfEvent and RawLink.
+type perfEventLink struct {
+	pe  *perfEvent
+	raw *RawLink
+}
+
+func (pl *perfEventLink) isLink() {}
+
+func (pl *perfEventLink) Pin(p string) error {
+	if pl.raw == nil {
+		return errors.New("perf link pin: perf event not attached")
+	}
+	return pl.raw.Pin(p)
+}
+
+func (pl *perfEventLink) Unpin() error {
+	if pl.raw == nil {
+		return errors.New("perf link unpin: perf event not attached")
+	}
+	return pl.raw.Unpin()
+}
+
+func (pl *perfEventLink) Close() error {
+	if pl.pe != nil {
+		if err := pl.pe.Close(); err != nil {
+			return fmt.Errorf("perf link close: %w", err)
+		}
+	}
+	if pl.raw == nil {
+		return nil
+	}
+	return pl.raw.Close()
+}
+
+func (pl *perfEventLink) Update(prog *ebpf.Program) error {
+	return fmt.Errorf("perf link update: %w", ErrNotSupported)
+}
+
+func (pl *perfEventLink) Info() (*Info, error) {
+	if pl.raw == nil {
+		return nil, errors.New("perf link info: perf event not attached")
+	}
+	return pl.raw.Info()
+}
+
+// newPerfEvent returns either perfEvent or perfEventLink, depending on
+// the availability of bpf perf link.
+//
+// perfEvent and perfEventLink implements Link.
+//
+// In order to use the perf event, attachPerfEvent must be called.
```

Comment:
A Link represents an existing attachment, not a possible attachment. Requiring attachPerfEvent later violates this principle. We should remove `newPerfEvent`.